pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -257,96 +257,62 @@ class StateVectorKokkos final
-     * @param wires Wires to apply gate to.
-     * @param inverse Indicates whether to use adjoint of gate.
-     */
-    void applyTwoQubitOp(const KokkosVector &matrix,
-                         const std::vector<size_t> &wires,
-                         bool inverse = false) {
-        auto &&num_qubits = this->getNumQubits();
-        if (!inverse) {
+        switch (wires.size()) {
+        case 1:
             Kokkos::parallel_for(
-                Kokkos::RangePolicy<KokkosExecSpace>(0, exp2(num_qubits - 2)),
-                twoQubitOpFunctor<fp_t, false>(*data_, num_qubits, matrix,
-                                               wires));
-        } else {
+                two2N, apply1QubitOpFunctor<fp_t>(*data_, num_qubits,
+                                                  matrix_trans, wires));
+            break;
+        case 2:
             Kokkos::parallel_for(
-                Kokkos::RangePolicy<KokkosExecSpace>(0, exp2(num_qubits - 2)),
-                twoQubitOpFunctor<fp_t, true>(*data_, num_qubits, matrix,
-                                              wires));
-        }
-    }
-
-    /**
-     * @brief Apply a multi qubit operator to the state vector using a matrix
-     *
-     * @param matrix Kokkos gate matrix in the device space
-     * @param wires Wires to apply gate to.
-     * @param inverse Indicates whether to use adjoint of gate.
-     */
-    void applyMultiQubitOp(const KokkosVector &matrix,
-                           const std::vector<size_t> &wires,
-                           bool inverse = false) {
-        auto &&num_qubits = this->getNumQubits();
-        if (wires.size() == 1) {
-            applySingleQubitOp(matrix, wires, inverse);
-        } else if (wires.size() == 2) {
-            applyTwoQubitOp(matrix, wires, inverse);
-        } else {
-            Kokkos::View<const size_t *, Kokkos::HostSpace,
-                         Kokkos::MemoryTraits<Kokkos::Unmanaged>>
-                wires_host(wires.data(), wires.size());
-
-            Kokkos::View<size_t *> wires_view("wires_view", wires.size());
-            Kokkos::deep_copy(wires_view, wires_host);
-
-            std::size_t two2N = std::exp2(num_qubits_ - wires.size());
-            std::size_t dim = std::exp2(wires.size());
+                two2N, apply2QubitOpFunctor<fp_t>(*data_, num_qubits,
+                                                  matrix_trans, wires));
+            break;
+        case 3:
+            Kokkos::parallel_for(
+                two2N, apply3QubitOpFunctor<fp_t>(*data_, num_qubits,
+                                                  matrix_trans, wires));
+            break;
+        case 4:
+            Kokkos::parallel_for(
+                two2N, apply4QubitOpFunctor<fp_t>(*data_, num_qubits,
+                                                  matrix_trans, wires));
+            break;
+        default:
             std::size_t scratch_size = ScratchViewComplex::shmem_size(dim) +
```

Comment:
Wondering if larger `shmem_size` can deliver better performance?

Reply:
AFAIK the size doesn't impact performance, it just has to be enough to store the `ScratchView`.